File: csharp/RIoT.Sdk.Facade/TaskClient.cs
using RIoT.Sdk.Core;

namespace RIoT.Sdk.Facade;

/// <summary>
/// Thin task-module facade. Expose additional endpoints as needed.
/// </summary>
public sealed class TaskClient
{
    private readonly RiotSession _session;

    internal TaskClient(RiotSession session) => _session = session;

    /// <summary>
    /// GET /api/task/vehicles/getAllVehicleSimpleInfo — list DispatchableVehicle rows (BC-VEH-002).
    /// Throws <see cref="RiotApiException"/> on business failure (ADR-sdk-0005).
    /// </summary>
    public async Task<IReadOnlyList<DispatchableVehicle>> GetDispatchableVehiclesAsync(
        CancellationToken cancellationToken = default)
    {
        var client = _session.CreateGeneratedTaskClient();
        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Task.Vehicles.GetAllVehicleSimpleInfo
                .GetAsync(cancellationToken: cancellationToken)
                .ConfigureAwait(false),
            "getAllVehicleSimpleInfo");

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);

        var result = response.Result ?? [];
        return result
            .Where(v => !string.IsNullOrWhiteSpace(v.DeviceKey) && !string.IsNullOrWhiteSpace(v.DeviceName))
            .Select(v => new DispatchableVehicle(v.DeviceKey!, v.DeviceName!))
            .ToList();
    }

    /// <summary>
    /// Exact deviceName → deviceKey via getAllVehicleSimpleInfo (BC-VEH-002).
    /// </summary>
    public async Task<string> ResolveDeviceKeyAsync(
        string deviceName,
        CancellationToken cancellationToken = default)
    {
        var vehicles = await GetDispatchableVehiclesAsync(cancellationToken).ConfigureAwait(false);
        return DispatchableVehicleLookup.ResolveDeviceKey(vehicles, deviceName);
    }

    /// <summary>
    /// POST /api/task/v1/route/getRouteCostsBy — RouteCost for one vehicle to a station (BC-ROUTE-001).
    /// <c>CostsMm = -1</c> means unreachable and is returned, not thrown (ADR-sdk-0005).
    /// </summary>
    public async Task<RouteCost> GetRouteCostAsync(
        int mapId,
        int stationId,
        string deviceKey,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(deviceKey);

        var client = _session.CreateGeneratedTaskClient();
        var body = new RIoT.Sdk.Generated.TaskApi.Models.U7533u8bf7u8f66u8f86u5217u8868u5230u8fbeu7ad9u70b9u4ee3u4ef7
        {
            MapId = mapId,
            StationId = stationId,
            DeviceKeys = [deviceKey],
        };

        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Task.V1.Route.GetRouteCostsBy
                .PostAsync(body, cancellationToken: cancellationToken)
                .ConfigureAwait(false),
            "getRouteCostsBy");

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);

        var entry = response.Result?.DeviceCostsList?
            .FirstOrDefault(c => string.Equals(c.DeviceKey, deviceKey, StringComparison.Ordinal));

        if (entry?.Costs is null)
        {
            throw new RiotApiException(
                $"getRouteCostsBy returned no costs for deviceKey={deviceKey}.",
                businessCode: "route-cost-missing");
        }

        return new RouteCost(entry.Costs.Value);
    }

    /// <summary>
    /// POST /api/task/v1/route/queryNearEnd — nearest end Station among candidates (NearStationQuery).
    /// Returns stationId only (not path geometry). Throws on business failure (ADR-sdk-0005).
    /// </summary>
    public async Task<int> QueryNearestEndAsync(
        int mapId,
        int startStationId,
        IReadOnlyList<int> endStationIds,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(endStationIds);
        if (endStationIds.Count == 0)
        {
            throw new ArgumentException("endStationIds must not be empty.", nameof(endStationIds));
        }

        var client = _session.CreateGeneratedTaskClient();
        var body = new RIoT.Sdk.Generated.TaskApi.Models.U627eu6700u8fd1u7684u7ec8u70b9u53c2u6570Object
        {
            MapId = mapId,
            StartStationId = startStationId,
            EndStationIds = endStationIds.Select(id => (int?)id).ToList(),
        };

        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Task.V1.Route.QueryNearEnd
                .PostAsync(body, cancellationToken: cancellationToken)
                .ConfigureAwait(false),
            "queryNearEnd");

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);

        return RiotBusinessResponse.RequireResult(
            response.Result,
            "queryNearEnd",
            "near-end-missing");
    }

    /// <summary>
    /// POST /api/task/v1/route/queryNearestStart — nearest start Station among candidates (NearStationQuery).
    /// Returns stationId only (not path geometry). Throws on business failure (ADR-sdk-0005).
    /// </summary>
    public async Task<int> QueryNearestStartAsync(
        int mapId,
        int endStationId,
        IReadOnlyList<int> startStationIds,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(startStationIds);
        if (startStationIds.Count == 0)
        {
            throw new ArgumentException("startStationIds must not be empty.", nameof(startStationIds));
        }

        var client = _session.CreateGeneratedTaskClient();
        var body = new RIoT.Sdk.Generated.TaskApi.Models.U627eu6700u8fd1u7684u8d77u70b9u53c2u6570Object
        {
            MapId = mapId,
            EndStationId = endStationId,
            StartStationIds = startStationIds.Select(id => (int?)id).ToList(),
        };

        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Task.V1.Route.QueryNearestStart
                .PostAsync(body, cancellationToken: cancellationToken)
                .ConfigureAwait(false),
            "queryNearestStart");

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);

        return RiotBusinessResponse.RequireResult(
            response.Result,
            "queryNearestStart",
            "near-start-missing");
    }

    /// <summary>
    /// POST /api/task/v1/order/command/{orderId} with CMD_ORDER_CANCEL (BC-ORDER-003).
    /// Uses string orderId; does not disable the vehicle. Throws on business failure (ADR-sdk-0005).
    /// </summary>
    public Task CancelOrderAsync(
        string orderId,
        string? reason = null,
        CancellationToken cancellationToken = default)
        => PostOrderCommandAsync(
            orderId,
            RIoT.Sdk.Generated.TaskApi.Models.OrderCommandDTOObject_commandType.CMD_ORDER_CANCEL,
            reason,
            cancellationToken);

    /// <summary>
    /// POST /api/task/v1/order/command/{orderId} with CMD_ORDER_HELD (BC-ORDER-006).
    /// Pauses an executing move order (orderState → HELD). Pair with OrderContinue.
    /// </summary>
    public Task OrderHoldAsync(
        string orderId,
        string? reason = null,
        CancellationToken cancellationToken = default)
        => PostOrderCommandAsync(
            orderId,
            RIoT.Sdk.Generated.TaskApi.Models.OrderCommandDTOObject_commandType.CMD_ORDER_HELD,
            reason,
            cancellationToken);

    /// <summary>
    /// POST /api/task/v1/order/command/{orderId} with CMD_ORDER_CONTINUE_FROM_HELD (BC-ORDER-006).
    /// Resumes a HELD move order. Do not use for OrderHang (use HangContinue).
    /// </summary>
    public Task OrderContinueAsync(
        string orderId,
        string? reason = null,
        CancellationToken cancellationToken = default)
        => PostOrderCommandAsync(
            orderId,
            RIoT.Sdk.Generated.TaskApi.Models.OrderCommandDTOObject_commandType.CMD_ORDER_CONTINUE_FROM_HELD,
            reason,
            cancellationToken);

    /// <summary>
    /// POST /api/task/v1/order/command/{orderId} with CMD_ORDER_CONTINUE_FROM_HANG (BC-ORDER-015).
    /// Attempts to pull an OrderHang back to EXECUTING. Do not use for HELD (use OrderContinue).
    /// HTTP/business success does not prove the order left HANG — confirm via orderState.
    /// </summary>
    public Task HangContinueAsync(
        string orderId,
        string? reason = null,
        CancellationToken cancellationToken = default)
        => PostOrderCommandAsync(
            orderId,
            RIoT.Sdk.Generated.TaskApi.Models.OrderCommandDTOObject_commandType.CMD_ORDER_CONTINUE_FROM_HANG,
            reason,
            cancellationToken);

    /// <summary>
    /// POST /api/task/vehicles/updateVehicleIntegrationLevel with serviceId=enable (BC-VEH-003).
    /// Brings the vehicle into dispatch (ON_LINE). Never pass ON_LINE as serviceId.
    /// </summary>
    public Task DispatchEnableAsync(
        string deviceKey,
        CancellationToken cancellationToken = default)
        => UpdateIntegrationLevelAsync(deviceKey, enable: true, cancellationToken);

    /// <summary>
    /// POST /api/task/vehicles/updateVehicleIntegrationLevel with serviceId=disable (BC-VEH-003).
    /// Removes the vehicle from dispatch (OFF_LINE). Does not cancel an executing order.
    /// </summary>
    public Task DispatchDisableAsync(
        string deviceKey,
        CancellationToken cancellationToken = default)
        => UpdateIntegrationLevelAsync(deviceKey, enable: false, cancellationToken);

    /// <summary>
    /// Access the underlying Kiota client for endpoints not yet wrapped.
    /// Prefer adding explicit facade methods for commonly used calls.
    /// </summary>
    public RIoT.Sdk.Generated.TaskApi.TaskClient Raw => _session.CreateGeneratedTaskClient();

    private async Task PostOrderCommandAsync(
        string orderId,
        RIoT.Sdk.Generated.TaskApi.Models.OrderCommandDTOObject_commandType commandType,
        string? reason,
        CancellationToken cancellationToken)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(orderId);

        var client = _session.CreateGeneratedTaskClient();
        var body = new RIoT.Sdk.Generated.TaskApi.Models.OrderCommandDTOObject
        {
            CommandType = commandType,
            DisableVehicle = false,
            Reason = reason,
        };

        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Task.V1.Order.Command[orderId]
                .PostAsync(body, cancellationToken: cancellationToken)
                .ConfigureAwait(false),
            commandType.ToString());

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);
    }

    private async Task UpdateIntegrationLevelAsync(
        string deviceKey,
        bool enable,
        CancellationToken cancellationToken)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(deviceKey);

        var serviceId = enable ? "enable" : "disable";
        var operation = enable ? "DispatchEnable" : "DispatchDisable";
        var client = _session.CreateGeneratedTaskClient();
        var body = new RIoT.Sdk.Generated.TaskApi.Models.BatchVehicleOperation
        {
            DeviceKeys = [deviceKey],
            ServiceId = serviceId,
        };

        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Task.Vehicles.UpdateVehicleIntegrationLevel
                .PostAsync(body, cancellationToken: cancellationToken)
                .ConfigureAwait(false),
            operation);

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);
    }
}

File: python/riot_sdk/facade/task_client.py
from __future__ import annotations

from typing import TYPE_CHECKING

from riot_sdk.core.business_response import (
    ensure_success,
    require_response,
    require_result,
)
from riot_sdk.core.dispatchable_vehicles import DispatchableVehicle, resolve_device_key
from riot_sdk.core.exceptions import RiotApiException
from riot_sdk.core.route_cost import RouteCost
from riot_sdk.generated.task.models.batch_vehicle_operation import BatchVehicleOperation
from riot_sdk.generated.task.models.order_command_d_t_o_object import OrderCommandDTOObject
from riot_sdk.generated.task.models.order_command_d_t_o_object_command_type import (
    OrderCommandDTOObject_commandType,
)
from riot_sdk.generated.task.models.u627eu6700u8fd1u7684u7ec8u70b9u53c2u6570_object import (
    U627eu6700u8fd1u7684u7ec8u70b9u53c2u6570Object,
)
from riot_sdk.generated.task.models.u627eu6700u8fd1u7684u8d77u70b9u53c2u6570_object import (
    U627eu6700u8fd1u7684u8d77u70b9u53c2u6570Object,
)
from riot_sdk.generated.task.models.u7533u8bf7u8f66u8f86u5217u8868u5230u8fbeu7ad9u70b9u4ee3u4ef7 import (
    U7533u8bf7u8f66u8f86u5217u8868u5230u8fbeu7ad9u70b9u4ee3u4ef7,
)

if TYPE_CHECKING:
    from riot_sdk.facade.session import RiotSession


class TaskClient:
    """Thin task-module facade."""

    def __init__(self, session: RiotSession) -> None:
        self._session = session

    async def get_dispatchable_vehicles(self) -> list[DispatchableVehicle]:
        """GET /api/task/vehicles/getAllVehicleSimpleInfo (BC-VEH-002 / ADR-sdk-0005)."""
        client = self._session.create_generated_task_client()
        response = require_response(
            await client.api.task.vehicles.get_all_vehicle_simple_info.get(),
            "getAllVehicleSimpleInfo",
        )
        ensure_success(response.code, response.message)
        result = response.result or []
        return [
            DispatchableVehicle(device_key=v.device_key, device_name=v.device_name)
            for v in result
            if v.device_key and v.device_name
        ]

    async def resolve_device_key(self, device_name: str) -> str:
        """Exact deviceName → deviceKey via getAllVehicleSimpleInfo (BC-VEH-002)."""
        vehicles = await self.get_dispatchable_vehicles()
        return resolve_device_key(vehicles, device_name)

    async def get_route_cost(
        self,
        map_id: int,
        station_id: int,
        device_key: str,
    ) -> RouteCost:
        """POST /api/task/v1/route/getRouteCostsBy (BC-ROUTE-001 / ADR-sdk-0005)."""
        if not device_key or not device_key.strip():
            raise ValueError("device_key is required")

        client = self._session.create_generated_task_client()
        body = U7533u8bf7u8f66u8f86u5217u8868u5230u8fbeu7ad9u70b9u4ee3u4ef7(
            map_id=map_id,
            station_id=station_id,
            device_keys=[device_key],
        )
        response = require_response(
            await client.api.task.v1.route.get_route_costs_by.post(body),
            "getRouteCostsBy",
        )
        ensure_success(response.code, response.message)

        costs_list = (response.result.device_costs_list if response.result else None) or []
        entry = next((c for c in costs_list if c.device_key == device_key), None)
        if entry is None or entry.costs is None:
            raise RiotApiException(
                f"getRouteCostsBy returned no costs for deviceKey={device_key}.",
                business_code="route-cost-missing",
            )
        return RouteCost(costs_mm=int(entry.costs))

    async def query_nearest_end(
        self,
        map_id: int,
        start_station_id: int,
        end_station_ids: list[int],
    ) -> int:
        """POST /api/task/v1/route/queryNearEnd — nearest end stationId (NearStationQuery)."""
        if not end_station_ids:
            raise ValueError("end_station_ids must not be empty")

        client = self._session.create_generated_task_client()
        body = U627eu6700u8fd1u7684u7ec8u70b9u53c2u6570Object(
            map_id=map_id,
            start_station_id=start_station_id,
            end_station_ids=list(end_station_ids),
        )
        response = require_response(
            await client.api.task.v1.route.query_near_end.post(body),
            "queryNearEnd",
        )
        ensure_success(response.code, response.message)
        return int(
            require_result(response.result, "queryNearEnd", "near-end-missing")
        )

    async def query_nearest_start(
        self,
        map_id: int,
        end_station_id: int,
        start_station_ids: list[int],
    ) -> int:
        """POST /api/task/v1/route/queryNearestStart — nearest start stationId (NearStationQuery)."""
        if not start_station_ids:
            raise ValueError("start_station_ids must not be empty")

        client = self._session.create_generated_task_client()
        body = U627eu6700u8fd1u7684u8d77u70b9u53c2u6570Object(
            map_id=map_id,
            end_station_id=end_station_id,
            start_station_ids=list(start_station_ids),
        )
        response = require_response(
            await client.api.task.v1.route.query_nearest_start.post(body),
            "queryNearestStart",
        )
        ensure_success(response.code, response.message)
        return int(
            require_result(response.result, "queryNearestStart", "near-start-missing")
        )

    async def cancel_order(self, order_id: str, reason: str | None = None) -> None:
        """POST /api/task/v1/order/command/{orderId} CMD_ORDER_CANCEL (BC-ORDER-003)."""
        await self._post_order_command(
            order_id,
            OrderCommandDTOObject_commandType.CMD_ORDER_CANCEL,
            reason,
        )

    async def order_hold(self, order_id: str, reason: str | None = None) -> None:
        """POST /api/task/v1/order/command/{orderId} CMD_ORDER_HELD (BC-ORDER-006)."""
        await self._post_order_command(
            order_id,
            OrderCommandDTOObject_commandType.CMD_ORDER_HELD,
            reason,
        )

    async def order_continue(self, order_id: str, reason: str | None = None) -> None:
        """POST /api/task/v1/order/command/{orderId} CMD_ORDER_CONTINUE_FROM_HELD (BC-ORDER-006)."""
        await self._post_order_command(
            order_id,
            OrderCommandDTOObject_commandType.CMD_ORDER_CONTINUE_FROM_HELD,
            reason,
        )

    async def hang_continue(self, order_id: str, reason: str | None = None) -> None:
        """POST /api/task/v1/order/command/{orderId} CMD_ORDER_CONTINUE_FROM_HANG (BC-ORDER-015)."""
        await self._post_order_command(
            order_id,
            OrderCommandDTOObject_commandType.CMD_ORDER_CONTINUE_FROM_HANG,
            reason,
        )

    async def dispatch_enable(self, device_key: str) -> None:
        """POST updateVehicleIntegrationLevel serviceId=enable (BC-VEH-003)."""
        await self._update_integration_level(device_key, enable=True)

    async def dispatch_disable(self, device_key: str) -> None:
        """POST updateVehicleIntegrationLevel serviceId=disable (BC-VEH-003)."""
        await self._update_integration_level(device_key, enable=False)

    @property
    def raw(self):
        """Underlying Kiota client for endpoints not yet wrapped."""
        return self._session.create_generated_task_client()

    async def _post_order_command(
        self,
        order_id: str,
        command_type: OrderCommandDTOObject_commandType,
        reason: str | None,
    ) -> None:
        if not order_id or not order_id.strip():
            raise ValueError("order_id is required")

        client = self._session.create_generated_task_client()
        body = OrderCommandDTOObject(
            command_type=command_type,
            disable_vehicle=False,
            reason=reason,
        )
        response = require_response(
            await client.api.task.v1.order.command.by_order_key(order_id).post(body),
            str(command_type.value),
        )
        ensure_success(response.code, response.message)

    async def _update_integration_level(self, device_key: str, *, enable: bool) -> None:
        if not device_key or not device_key.strip():
            raise ValueError("device_key is required")

        service_id = "enable" if enable else "disable"
        operation = "DispatchEnable" if enable else "DispatchDisable"
        client = self._session.create_generated_task_client()
        body = BatchVehicleOperation(
            device_keys=[device_key],
            service_id=service_id,
        )
        response = require_response(
            await client.api.task.vehicles.update_vehicle_integration_level.post(body),
            operation,
        )
        ensure_success(response.code, response.message)
